src/apps/fuzzers/fuzzerPolygonToCells.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2022 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/** @file
+ * @brief Fuzzer program for polygonToCells and related functions
+ */
+
+#include "aflHarness.h"
+#include "h3api.h"
+#include "utility.h"
+
+typedef struct {
+    int res;
+    int numHoles;
+    // repeating: num verts, verts
+} inputArgs;
+
+const int MAX_RES = 15;
+const int MAX_SZ = 4000000;
+const int MAX_HOLES = 100;
+
+int populateGeoLoop(GeoLoop *g, const uint8_t *data, size_t *offset,
+                    size_t size) {
+    if (size < *offset + sizeof(int)) {
+        return 1;
+    }
+    int numVerts = *(const int *)(data + *offset);
+    *offset = *offset + sizeof(int);
+    g->numVerts = numVerts;
+    if (size < *offset + sizeof(LatLng) * numVerts) {
+        return 1;
+    }
+    g->verts = (LatLng *)(data + *offset);
+    *offset = *offset + sizeof(LatLng) * numVerts;
+    return 0;
+}
+
+void run(GeoPolygon *geoPolygon, int res) {
+    int64_t sz;
+    H3Error err = H3_EXPORT(maxPolygonToCellsSize)(geoPolygon, res, &sz);
+    if (!err && sz < MAX_SZ) {
+        if (sz < 0) {
+            printf("Oh no - sz is negative\n");
+        }
+        H3Index *out = calloc(sz, sizeof(H3Index));
+        H3_EXPORT(polygonToCells)(geoPolygon, res, out);
+        free(out);
+    }
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+    if (size < sizeof(inputArgs)) {
+        return 0;
+    }
+    const inputArgs *args = (const inputArgs *)data;
+    int res = args->res % (MAX_RES + 1);
+
+    GeoPolygon geoPolygon;
+    geoPolygon.numHoles = args->numHoles % MAX_HOLES;
+    if (geoPolygon.numHoles < 0) {
+        return 0;
+    }
+    geoPolygon.holes = calloc(geoPolygon.numHoles, sizeof(GeoLoop));
+    size_t offset = sizeof(inputArgs);
```

Comment:
I'm quite confused here - isn't the fuzzer only going to provide `sizeof(inputArgs)` worth of data? Where does the rest of the data after `offset` come from in that case?

Reply:
I will need to check if this will work with AFL (might need to adjust the seed file so it has data after inputArgs). For libFuzzer it does not know about `inputArgs` so it will generate larger test cases.